### DISCUSSION TITLE
Issue with Traffic Forwarding when Multiple Road Warriors Connect to StrongSwan Server

### DISCUSSION THREAD

**Discussion by Wonder2018** [strongswan/strongswan#2574] [Q&A] [answered]
> Is there a problem with my configuration? Or is it a bug?

**System (please complete the following information):**

- OS: Ubuntu 20.04
- Kernel version (if applicable): 5.15.0-71-generic
- strongSwan version(s): I have tried both 5.9 and 6.0
- Tested/confirmed with the latest version: yes

**Describe the bug**

When two road warriors (clients) are simultaneously connected to a StrongSwan server, the server encounters issues with traffic forwarding.

**To Reproduce**

1. Both the server and clients are behind NAT, but the server's UDP ports 500 and 4500 are mapped to a public IP.
2. Generate a CA certificate.
3. Use the CA certificate to generate a server certificate and two sets of client certificates.
4. Start the server using the provided swanctl configuration.
5. Add the provided iptables rules.
6. Create IKEv2 connections on two Windows machines and connect to the server using the generated client certificates.
7. Ping devices within the local_ts range from both Windows systems.
8. Use tcpdump 'udp port 500 or udp port 4500' on the server to observe the traffic.

**What happened**

1. At any given time, only one Windows client receives ping replies.
2. tcpdump shows that the server generates ping replies, but these replies are all sent to one client, regardless of which client originated the ping packet.
3. After some time, the clients may switch states, meaning that initially, all ping replies are sent to client A, but after a while, all ping replies are sent to client B.
4. The client that established the connection last receives all ping replies.

**Expected behavior**
Both clients should be able to establish connections to the server simultaneously and receive their respective ping replies normally.

**Logs/Backtraces**

tcpdump output: 

```log
root@strongswan:/# tcpdump 'udp port 500 or udp port 4500'
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
14:17:30.855038 IP x.x.x.x.15144 > strongswan.ipsec-nat-t: UDP-encap: ESP(spi=0xc0e00e06,seq=0xc), length 100
14:17:30.855156 IP strongswan.ipsec-nat-t > x.x.x.x.15144: UDP-encap: ESP(spi=0xd3cc8d66,seq=0xd), length 100
14:17:31.873825 IP x.x.x.x.15144 > strongswan.ipsec-nat-t: UDP-encap: ESP(spi=0xc0e00e06,seq=0xd), length 100
14:17:31.873945 IP strongswan.ipsec-nat-t > x.x.x.x.15144: UDP-encap: ESP(spi=0xd3cc8d66,seq=0xe), length 100
14:17:32.895059 IP x.x.x.x.15144 > strongswan.ipsec-nat-t: UDP-encap: ESP(spi=0xc0e00e06,seq=0xe), length 100
14:17:32.895181 IP strongswan.ipsec-nat-t > x.x.x.x.15144: UDP-encap: ESP(spi=0xd3cc8d66,seq=0xf), length 100
14:17:33.083940 IP x.x.x.x.15144 > strongswan.ipsec-nat-t: isakmp-nat-keep-alive
14:17:33.908221 IP x.x.x.x.15144 > strongswan.ipsec-nat-t: UDP-encap: ESP(spi=0xc0e00e06,seq=0xf), length 100
14:17:33.908411 IP strongswan.ipsec-nat-t > x.x.x.x.15144: UDP-encap: ESP(spi=0xd3cc8d66,seq=0x10), length 100
14:17:34.930951 IP x.x.x.x.15144 > strongswan.ipsec-nat-t: UDP-encap: ESP(spi=0xc0e00e06,seq=0x10), length 100
14:17:34.931080 IP strongswan.ipsec-nat-t > x.x.x.x.15144: UDP-encap: ESP(spi=0xd3cc8d66,seq=0x11), length 100

14:17:38.611140 IP x.x.x.x.15159 > strongswan.ipsec-nat-t: isakmp-nat-keep-alive
14:17:40.046176 IP x.x.x.x.15159 > strongswan.ipsec-nat-t: UDP-encap: ESP(spi=0xc59aeca6,seq=0x77), length 100
14:17:40.046305 IP strongswan.ipsec-nat-t > x.x.x.x.15144: UDP-encap: ESP(spi=0xd3cc8d66,seq=0x12), length 100
14:17:41.543172 IP x.x.x.x.15159 > strongswan.ipsec-nat-t: UDP-encap: ESP(spi=0xc59aeca6,seq=0x78), length 100
14:17:41.543294 IP strongswan.ipsec-nat-t > x.x.x.x.15144: UDP-encap: ESP(spi=0xd3cc8d66,seq=0x13), length 100
14:17:43.042681 IP x.x.x.x.15159 > strongswan.ipsec-nat-t: UDP-encap: ESP(spi=0xc59aeca6,seq=0x79), length 100
14:17:43.042802 IP strongswan.ipsec-nat-t > x.x.x.x.15144: UDP-encap: ESP(spi=0xd3cc8d66,seq=0x14), length 100
14:17:44.535701 IP x.x.x.x.15159 > strongswan.ipsec-nat-t: UDP-encap: ESP(spi=0xc59aeca6,seq=0x7a), length 100
14:17:44.535832 IP strongswan.ipsec-nat-t > x.x.x.x.15144: UDP-encap: ESP(spi=0xd3cc8d66,seq=0x15), length 100
```

**Additional context**

**compile with**

```bash
 ./configure --prefix=/usr --sysconfdir=/etc --disable-defaults      \
    --enable-charon --enable-ikev2 --enable-nonce --enable-random     \
    --enable-openssl --enable-pem --enable-x509 --enable-pubkey       \
    --enable-constraints --enable-pki --enable-socket-default         \
    --enable-kernel-netlink --enable-swanctl --enable-resolve         \
    --enable-eap-identity --enable-eap-md5 --enable-eap-dynamic       \
    --enable-eap-tls --enable-updown --enable-vici --enable-ml        \
    --enable-silent-rules  && \
   make all && make install
```

**swanctl.conf**

```conf
connections {
  myecs {

    version = 2
    unique = never
    rekey_time = 0s
    dpd_delay = 60s
    proposals = aes256-sha256-modp1024,aes128-sha256-modp2048
    pools = primary-pool-ipv4
    dpd_delay = 60s
    encap = yes

    local {
      auth = pubkey
      certs = server.pem
      id = wonderland.inc
    }

    remote {
      auth = eap-tls
      eap_id = %any
      cacerts = myecs.pem
    }

    children {

      child-net {

        local_ts = 10.1.100.0/24, 10.1.101.0/24
        remote_ts = 10.1.200.0/24

        esp_proposals = aes256-sha1,aes128-sha1,aes256-sha256,aes128-sha256

        dpd_action = clear
        start_action = trap
        close_action = clear
        rekey_time = 0s
      }
    }
  }
}

pools {
  primary-pool-ipv4 {
    addrs =10.1.200.0/24
  }
}
```

**/etc/strongswan.d/charon-logging.conf**

```conf
charon {
   start-scripts {
      creds = swanctl --load-creds
      conns = swanctl --load-conns
      pools = swanctl --load-pools
   }
   filelog {
      stderr {
         default = 1
      }
   }
   eap-dynamic {
      prefer_user = yes
      preferred = md5, tls
   }

}

libtls {
  version_max = 1.3
  suites = TLS_AES_128_GCM_SHA256,TLS_AES_256_GCM_SHA384
}

```

**iptables**

```bash
iptables -t nat -A POSTROUTING -s 10.1.200.0/24 ! -d 10.1.200.0/24 -j MASQUERADE
```

**Comment by tobiasbrunner** [comment, marked as answer]
> remote_ts = 10.1.200.0/24

This is wrong when using virtual IPs, remove it (see [the docs](https://docs.strongswan.org/docs/latest/features/vip.html#_responder_configuration)).

**Reply by Wonder2018** [reply]
## Thank you for your answer! 

I tried removing `remote_ts = 10.1.200.0/24` from the configuration, but the same issue still persisted. 

However, when I changed the configuration to `remote_ts = dynamic`, the problem no longer occurred, and everything ran as expected. 

Thanks again for your help.

**Reply by tobiasbrunner** [reply]
Weird, that's the default value if the option is not configured.

**Reply by Wonder2018** [reply]
I tried resetting the network configurations of all devices and then removed `remote_ts = 10.1.200.0/24` from swanctl.conf. The server indeed worked as expected. Perhaps some extraneous iptables entries or other incorrect network configurations caused this issue.

I'm very sorry to have confused you
